### PR TITLE
Add function to get version as of a certain date. Also formatting

### PR DESCRIPTION
**Pull request by changhiskhan**

### CODE REVIEW

File: python/lance/__init__.py [hunk 1]
```diff
@@ -68,15 +76,26 @@ def dataset(
...
             return _dataset_plain(uri, filesystem=filesystem, **kwargs)
 
+        if asof is not None:
+            ds = dataset(uri, version, filesystem=filesystem, **kwargs)
+            version = get_version_asof(ds, asof)
+
+    return _get_versioned_dataset(filesystem, uri, version)
+
+
+def _is_versioned(filesystem, uri):
```
Comment (eddyxu) [inline python/lance/__init__.py:93]: not sure what does this mean. it checks whether latest manifest exist? why it is related to versioned?

Or you mean it is a plain dataset? Should we just rename this function to `is_plain_dataset`?

File: python/lance/__init__.py [hunk 2]
```diff
@@ -68,15 +76,26 @@ def dataset(
...
+    manifest = os.path.join(uri, "_latest.manifest")
+    return not (filesystem.get_file_info(manifest).type == pa.fs.FileType.NotFound)
+
+
+def _get_versioned_dataset(filesystem, uri, version):
```
Comment (eddyxu) [inline python/lance/__init__.py:98]: let's add type annotations here?

File: python/lance/__init__.py [hunk 3]
```diff
@@ -68,15 +76,26 @@ def dataset(
...
     if version is None:
-        if (
-            filesystem.get_file_info(os.path.join(uri, "_latest.manifest")).type
-            == pa.fs.FileType.NotFound
-        ):
+        if not _is_versioned(filesystem, uri):
             return _dataset_plain(uri, filesystem=filesystem, **kwargs)
 
+        if asof is not None:
+            ds = dataset(uri, version, filesystem=filesystem, **kwargs)
```
Comment (eddyxu) [inline python/lance/__init__.py:87]: use `_get_versioned_dataset` here? to reduce another I/O to check `is_plain_dataset` ?

File: python/lance/__init__.py (hunk 1 above)
Comment (changhiskhan) [reply python/lance/__init__.py:93]: i'm just refactoring what you had in there

Comment (changhiskhan) [reply python/lance/__init__.py:93]: sure i can call this "is_plain_dataset"

File: python/lance/__init__.py (hunk 2 above)
Comment (changhiskhan) [reply python/lance/__init__.py]: done

File: python/lance/__init__.py (hunk 3 above)
Comment (changhiskhan) [reply python/lance/__init__.py]: done

File: python/lance/__init__.py (hunk 1 above)
Comment (changhiskhan) [reply python/lance/__init__.py]: done

**Review by eddyxu — APPROVED** [review]

**Pull request merged**